### PR TITLE
:seedling: Release v1.7.0

### PR DESCRIPTION
**Pull request by willie-yao**
/area release

### CODE REVIEW

**Comment by cahillsf** [comment]
seems like this is back to the older format -- is it intentional that we're moving away from the collapsible details?

File: CHANGELOG/v1.7.0.md [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+- util: Checking cert's keypair for nil before accessing to avoid panics (#10368)
+- util: recover gvk after scheme.Convert (#10409)
+
+## :seedling: Others
+- API: Stop relying on GVK being set on regular typed objects (#9956)
+- Bootstrap: Add MachinePool test for switching bootstrap config to another ready/non-ready object with different name (#9616)
+- CABPK: Add pod metadata to capbk manager (#10208)
+- CAPD: make docker machine bootstrap timeout configurable (#9952)
+- CAPD: Remove requeues in DockerMachinePool (#9725)
+- CAPD:: Support configuring ExtraPortMappings for the kind cluster (#10046)
```
Comment (typeid) [inline CHANGELOG/v1.7.0.md:103]:
```suggestion
- CAPD: Support configuring ExtraPortMappings for the kind cluster (#10046)
```

File: CHANGELOG/v1.7.0.md [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+- CI: Bump kubebuilder envtest to 1.29.0 (#10013)
+- CI: Change base branch for link checker (#10365)
+- CI: DumpResourcesForCluster should fail fast for i/o errors (#10238)
+- CI: Ensure build images contains correct binary and for correct architecture (#9932)
+- CI: Fix Make target generate-go-openapi (#10161)
+- CI: Fix TestPatchNode flake (#10287)
+- CI: Hack/prowjob-gen skip creating empty files (#10022)
+- CI: implement generator for prowjobs (#9937)
+- CI: Makefile bump dependencies (#10236)
+- CI: bump WorkloadKubernetesVersion for v1.6 clusterctl upgrade test (#10017)
```
Comment (typeid) [inline CHANGELOG/v1.7.0.md:125]:
```suggestion
- CI: Bump WorkloadKubernetesVersion for v1.6 clusterctl upgrade test (#10017)
```

**Comment by sbueringer** [comment]
> seems like this is back to the older format -- is it intentional that we're moving away from the collapsible details?

I'm not sure, but was this maybe just for beta / RC? So we have the delta since last beta / RC directly and the full list collapsible?

File: CHANGELOG/v1.7.0.md [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+- Workload Cluster: v1.24.x -> v1.29.x
+
+[More information about version support can be found here](https://cluster-api.sigs.k8s.io/reference/versions.html)
+
+
+## Highlights
+
+* Set MachinePool feature flag default to true + Beta
+* Enable kubeadm feature gates mutation
+* Add public function to create new CRD migrator
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:13]: I wouldn't mention this as a highlight

File: CHANGELOG/v1.7.0.md [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+* Add MachineSetReady condition to MachineDeployment
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:14]: Same here, just too minor in my opinion

File: CHANGELOG/v1.7.0.md [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+- MachineHealthCheck: implement annotation to manually mark machines for remediation (#10202)
+- Release: Add dependencies section to release notes tool (#10043)
+- Release: Add release notes expander functionality (#10091)
+- Release: Add utility to create git issues on provider repo (#9110)
+- Testing: Resolve release markers (#9265)
+- Testing: Prepare quickstart, capd and tests for the new release including kind bump (#9872)
+
+## :bug: Bug Fixes
+- API: Use ptr instead of own implementation (#10276)
+- API: Restore v1alpha3/v1alpha4 conversion to fix SSA issue & add e2e test coverage (#10147)
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:49]: Let's drop `- API: Remove v1alpha4 API Version (#9939)` above

and add it here instead:

```
- API: Restore v1alpha3/v1alpha4 conversion to fix SSA issue & add e2e test coverage (#9939 #10147) (for more details see #10051)
```

File: CHANGELOG/v1.7.0.md [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+- 5 breaking changes :warning:
+- 12 feature additions ✨
+- 48 bugs fixed 🐛
+
+## :memo: Proposals
+- MachinePool: Update MachinePool Machines proposal with diagrams (#9664)
+
+## :warning: Breaking Changes
+- API: Remove v1alpha4 API Version (#9939)
+- Dependency: Bump to controller-runtime v0.17.0 & k8s.io v0.29 (#9964)
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:28]:
```suggestion
- Dependency: Bump to controller-runtime v0.17.0 & k8s.io/* v0.29 (#9964)
```

File: CHANGELOG/v1.7.0.md [hunk 7]
```diff
@@ -0,0 +1,433 @@
+## 👌 Kubernetes version support
+
+- Management Cluster: v1.26.x -> v1.29.x
+- Workload Cluster: v1.24.x -> v1.29.x
+
+[More information about version support can be found here](https://cluster-api.sigs.k8s.io/reference/versions.html)
+
+
+## Highlights
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:9]: Let's add the following instead:

* MachinePool: promoted to beta and enabled per default (#10141)
* MachineDeployment: Taint nodes with PreferNoSchedule during rollouts (#10223)
* MachineHealthCheck: implement annotation to manually mark machines for remediation (#10202)
* KCP improvements:
  * Enable kubeadm feature gates mutation (#10154)
  * Defer remediation when a control plane machine is still provisioning (#9734)
  * Prefer deletion of Machines with unhealthy control plane components during rollout (#10421)
* ClusterClass:
  * Add variable metadata (#10308)
  * Improved version validation if an update is already in progress (#10063)
  * Improved unit and e2e test coverage for ClusterClass with MachinePools (#10028)
* Various improvements to our e2e tests to increase test coverage and reduce flakes to improve our CI signal
* Implemented prowjob-gen to make it easier to manage our ProwJobs (#9937)

File: CHANGELOG/v1.7.0.md [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+- 339 new commits merged
+- 5 breaking changes :warning:
+- 12 feature additions ✨
+- 48 bugs fixed 🐛
+
+## :memo: Proposals
+- MachinePool: Update MachinePool Machines proposal with diagrams (#9664)
+
+## :warning: Breaking Changes
+- API: Remove v1alpha4 API Version (#9939)
```
Comment (chrischdi) [inline CHANGELOG/v1.7.0.md:27]: This was partially reverted in (below line 49): 
```
- API: Restore v1alpha3/v1alpha4 conversion to fix SSA issue & add e2e test coverage (#10147)
```

However, the restore is a bit different from what we had before, it also restores v1alpha3 for conversion. 

With this release, v1alpha4 types cannot be imported anymore in external go modules (which still can be argued to be a breaking changes)

But @sbueringer may have the best take on this :-)

File: CHANGELOG/v1.7.0.md [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+- KCP/MachineSet: Objects generated by KCP, MachineSets and MachinePools will now consistently use machine name (#9833)
+
+## :sparkles: New Features
+- API: Enable kubeadm feature gates mutation (#10154)
+- clusterctl: Add k0smotron provider (#9991)
+- clusterctl: Add public function to create new CRD migrator (#10075)
+- Control-plane: Allow mutation of all fields that should be mutable (#9871)
+- MachineDeployment: Add MachineSetReady condition to MachineDeployment (#9262)
+- MachineDeployment: Taint nodes with PreferNoSchedule during rollouts (#10223)
+- MachineHealthCheck: implement annotation to manually mark machines for remediation (#10202)
```
Comment (chrischdi) [inline CHANGELOG/v1.7.0.md:40]: Not 100% sure if these qualify as highlight but these are nice features :-).

File: CHANGELOG/v1.7.0.md [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+- Dependency: Bump github.com/onsi/ginkgo/v2 from 2.16.0 to 2.17.0 (#10284)
+- Dependency: Bump github.com/onsi/gomega from 1.31.1 to 1.32.0 (#10285)
+- Dependency: Bump Go to 1.21.5 (#9900)
+- Dependency: Bump go version to 1.21.8 (#10235)
+- Dependency: Bump golang to v1.21.9 and golang.org/x/net to mitigate CVE-2023-45288 (#10378)
+- Dependency: Bump kind version to v0.22.0 (#10094)
+- Dependency: Bump protobuf to v1.33.0 to address CVEs (#10248)
+- Dependency: Bump sigs.k8s.io/controller-runtime from 0.17.2 to 0.17.3 (#10406)
+- Dependency: Bump the kubernetes group to v0.28.5 (#9933)
+- Dependency: Bump the kubernetes group with 8 updates (#10286)
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:163]:
```suggestion
```

File: CHANGELOG/v1.7.0.md [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+- clusterctl: Clarify rules for adding new clusterctl default providers (#9975)
+- clusterctl: Deprecate clusterctl alpha topology plan (#10139)
+- clusterctl: Replace context.TODO() from clusterctl proxy.go (#9776)
+- Community meeting: Move CecileRobertMichon to emeritus (#10042)
+- Community meeting: Move ykakarap to emeritus approvers (#10000)
+- Community meeting: Promote chrischdi to Cluster API maintainer (#9997)
+- Dependency: Bump `github.com/docker/docker` from 24.0.7 -> 25.0.0 (#10057)
+- Dependency: Bump `golangci-lint` to v1.56.1 (#10124)
+- Dependency: Bump controller-runtime to v0.17.1 (#10131)
+- Dependency: Bump controller-runtime to v0.17.2 (#10162)
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply CHANGELOG/v1.7.0.md:151]: I think we don't have to do this for every release, but it's nice to get rid of some noise on release notes of a minor version

File: CHANGELOG/v1.7.0.md [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+- Dependency: Bump controller-tools to v0.14 (#9987)
+- Dependency: Bump crypto dep version (#9938)
+- Dependency: Bump github.com/onsi/ginkgo/v2 from 2.16.0 to 2.17.0 (#10284)
+- Dependency: Bump github.com/onsi/gomega from 1.31.1 to 1.32.0 (#10285)
```
Comment (chrischdi) [inline CHANGELOG/v1.7.0.md:155]: optional: already part of below dependencies section? propably matches more of these `Dependency: *` parts?

File: CHANGELOG/v1.7.0.md [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+- Release: Improve weekly PR update generation script and documentation (#10092)
+- Release: Make release notes tool not dependent on local git (#9618)
+- Release: Prep main branch for v1.7 development (#9799)
+- Release: Provide faster support for new k8s minor releases (#9971)
+- Release: Read in dockerfiles from stdin (#9990)
+- Release: also detect alpha releases as pre releases (#10379)
+- Release: Release v1.5.7 (#10251)
+- Release: Release v1.6.3 (#10252)
+- Release: Swap in new 1.7 RT members in OWNERS (#9855)
+- Runtime SDK: Add more helper functions in topologymutation varaible.go to help unmarshal variables (#9670)
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:213]:
```suggestion
- Runtime SDK: Add more helper functions in topologymutation variable.go to help unmarshal variables (#9670)
```

File: CHANGELOG/v1.7.0.md [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+- Testing: Small improvements to the inmemory api server (#9935)
+- Testing: add dynamic finalizer assertions for e2e framework (#9970)
+- Testing: Watch for Cluster resources in topology MachineSet & MachineDeployment controllers (#10029)
+- Testing: Watch for Cluster resources in topology MD controller (#9865)
+- util: Improve patch helper error handling (#9946)
+- util: Use min/max funcs from Go SDK (#9945)
+
+:book: Additionally, there have been 53 contributions to our documentation and book. (#10005, #10031, #10040, #10061, #10066, #10068, #10084, #10099, #10100, #10115, #10122, #10170, #10174, #10194, #10239, #10257, #10268, #10288, #10289, #10323, #10329, #10334, #10381, #10393, #10410, #9585, #9640, #9767, #9771, #9779, #9782, #9786, #9794, #9797, #9801, #9817, #9829, #9831, #9838, #9856, #9866, #9867, #9868, #9876, #9896, #9897, #9908, #9941, #9949, #9957, #9961, #9972, #9993) 
+
+## Dependencies
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:231]: I wonder if we should mention that this is for our main go module (maybe a short sentence below?)

File: CHANGELOG/v1.7.0.md (hunk 8 above)
Comment (sbueringer) [reply CHANGELOG/v1.7.0.md:27]: I added a comment about this below

File: CHANGELOG/v1.7.0.md (hunk 9 above)
Comment (sbueringer) [reply CHANGELOG/v1.7.0.md:40]: Agree, also included in my highlight section proposal below

File: CHANGELOG/v1.7.0.md (hunk 12 above)
Comment (sbueringer) [reply CHANGELOG/v1.7.0.md:155]: Agree. Suggested to drop them below

**Comment by willie-yao** [comment]
> seems like this is back to the older format -- is it intentional that we're moving away from the collapsible details?

The collapsable details are for beta and RC only

File: CHANGELOG/v1.7.0.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (willie-yao) [reply CHANGELOG/v1.7.0.md:151]: I think these dependency PRs are only excluded if they are created by dependabot. This is by design as discussed here: https://github.com/kubernetes-sigs/cluster-api/pull/10043#discussion_r1467308837

File: CHANGELOG/v1.7.0.md (hunk 14 above)
Comment (willie-yao) [reply CHANGELOG/v1.7.0.md:231]: I just added `for the main go module` to the sentence. Would that make sense or is more info needed in a separate sentence?

Comment (sbueringer) [reply CHANGELOG/v1.7.0.md:228]: That's fine

File: CHANGELOG/v1.7.0.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply CHANGELOG/v1.7.0.md]: Yup. I remember. I think it's fine that we generate it like that. I think it just makes sense to clean up this list a bit manually for minor releases

File: CHANGELOG/v1.7.0.md [hunk 15]
```diff
@@ -0,0 +1,430 @@
...
+  * Enable kubeadm feature gates mutation (#10154)
+  * Defer remediation when a control plane machine is still provisioning (#9734)
+  * Prefer deletion of Machines with unhealthy control plane components during rollout (#10421)
+* ClusterClass:
+  * Add variable metadata (#10308)
+  * Improved version validation if an update is already in progress (#10063)
+  * Improved unit and e2e test coverage for ClusterClass with MachinePools (#10028)
+* Various improvements to our e2e tests to increase test coverage and reduce flakes to improve our CI signal
+* Implemented prowjob-gen to make it easier to manage our ProwJobs (#9937)
+* Enable kubeadm feature gates mutation (#10154)
```
Comment (sbueringer) [inline CHANGELOG/v1.7.0.md:24]:
```suggestion
```
(already part of KCP improvements)

File: CHANGELOG/v1.7.0.md (hunk 14 above)
Comment (sbueringer) [reply CHANGELOG/v1.7.0.md:228]: Oh wait. I meant to the dependencies section not to the book (sorry selected the wrong lines)

Comment (sbueringer) [reply CHANGELOG/v1.7.0.md:228]: If we can't easily add it to the Dependencies section because of the way that is generated, it's fine

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve

/hold

**Pull request merged**